// directus/dist/server.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const terminus_1 = require("@godaddy/terminus");
const http = __importStar(require("http"));
const https = __importStar(require("https"));
const lodash_1 = require("lodash");
const qs_1 = __importDefault(require("qs"));
const url_1 = __importDefault(require("url"));
const app_1 = __importDefault(require("./app"));
const database_1 = __importDefault(require("./database"));
const emitter_1 = require("./emitter");
const env_1 = __importDefault(require("./env"));
const logger_1 = __importDefault(require("./logger"));
async function createServer() {
    const server = http.createServer(await app_1.default());
    server.on('request', function (req, res) {
        const startTime = process.hrtime();
        const complete = lodash_1.once(function (finished) {
            var _a, _b, _c;
            const elapsedTime = process.hrtime(startTime);
            const elapsedNanoseconds = elapsedTime[0] * 1e9 + elapsedTime[1];
            const elapsedMilliseconds = elapsedNanoseconds / 1e6;
            const previousIn = ((_a = req.socket._metrics) === null || _a === void 0 ? void 0 : _a.in) || 0;
            const previousOut = ((_b = req.socket._metrics) === null || _b === void 0 ? void 0 : _b.out) || 0;
            const metrics = {
                in: req.socket.bytesRead - previousIn,
                out: req.socket.bytesWritten - previousOut,
            };
            req.socket._metrics = {
                in: req.socket.bytesRead,
                out: req.socket.bytesWritten,
            };
            // Compatibility when supporting serving with certificates
            const protocol = server instanceof https.Server ? 'https' : 'http';
            // Rely on url.parse for path extraction
            // Doesn't break on illegal URLs
            const urlInfo = url_1.default.parse(req.originalUrl || req.url);
            const info = {
                finished,
                request: {
                    aborted: req.aborted,
                    completed: req.complete,
                    method: req.method,
                    url: urlInfo.href,
                    path: urlInfo.pathname,
                    protocol,
                    host: req.headers.host,
                    size: metrics.in,
                    query: urlInfo.query ? qs_1.default.parse(urlInfo.query) : {},
                    headers: req.headers,
                },
                response: {
                    status: res.statusCode,
                    size: metrics.out,
                    headers: res.getHeaders(),
                },
                ip: req.headers['x-forwarded-for'] || ((_c = req.socket) === null || _c === void 0 ? void 0 : _c.remoteAddress),
                duration: elapsedMilliseconds.toFixed(),
            };
            emitter_1.emitAsyncSafe('response', info);
        });
        res.once('finish', complete.bind(null, true));
        res.once('close', complete.bind(null, false));
    });
    const terminusOptions = {
        timeout: 1000,
        signals: ['SIGINT', 'SIGTERM', 'SIGHUP'],
        beforeShutdown,
        onSignal,
        onShutdown,
    };
    terminus_1.createTerminus(server, terminusOptions);
    return server;
    async function beforeShutdown() {
        emitter_1.emitAsyncSafe('server.stop.before', { server });
        if (env_1.default.NODE_ENV !== 'development') {
            logger_1.default.info('Shutting down...');
        }
    }
    async function onSignal() {
        const database = database_1.default();
        await database.destroy();
        logger_1.default.info('Database connections destroyed');
    }
    async function onShutdown() {
        emitter_1.emitAsyncSafe('server.stop');
        if (env_1.default.NODE_ENV !== 'development') {
            logger_1.default.info('Directus shut down OK. Bye bye!');
        }
    }
}
exports.default = createServer;
