// directus/dist/cli/commands/init/index.js
"use strict";
/* eslint-disable no-console */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const argon2_1 = __importDefault(require("argon2"));
const chalk_1 = __importDefault(require("chalk"));
const execa_1 = __importDefault(require("execa"));
const inquirer_1 = __importDefault(require("inquirer"));
const ora_1 = __importDefault(require("ora"));
const uuid_1 = require("uuid");
const run_1 = __importDefault(require("../../../database/migrations/run"));
const run_2 = __importDefault(require("../../../database/seeds/run"));
const create_db_connection_1 = __importDefault(require("../../utils/create-db-connection"));
const create_env_1 = __importDefault(require("../../utils/create-env"));
const drivers_1 = require("../../utils/drivers");
const questions_1 = require("./questions");
async function init() {
    const rootPath = process.cwd();
    const { client } = await inquirer_1.default.prompt([
        {
            type: 'list',
            name: 'client',
            message: 'Choose your database client',
            choices: Object.values(drivers_1.drivers),
        },
    ]);
    const dbClient = drivers_1.getDriverForClient(client);
    const spinnerDriver = ora_1.default('Installing Database Driver...').start();
    await execa_1.default('npm', ['install', dbClient, '--production']);
    spinnerDriver.stop();
    let attemptsRemaining = 5;
    const { credentials, db } = await trySeed();
    async function trySeed() {
        const credentials = await inquirer_1.default.prompt(questions_1.databaseQuestions[dbClient].map((question) => question({ client: dbClient, filepath: rootPath })));
        const db = create_db_connection_1.default(dbClient, credentials);
        try {
            await run_2.default(db);
            await run_1.default(db, 'latest');
        }
        catch (err) {
            console.log();
            console.log('Something went wrong while seeding the database:');
            console.log();
            console.log(`${chalk_1.default.red(`[${err.code || 'Error'}]`)} ${err.message}`);
            console.log();
            console.log('Please try again');
            console.log();
            attemptsRemaining--;
            if (attemptsRemaining > 0) {
                return await trySeed();
            }
            else {
                console.log(`Couldn't seed the database. Exiting.`);
                process.exit(1);
            }
        }
        return { credentials, db };
    }
    await create_env_1.default(dbClient, credentials, rootPath);
    console.log();
    console.log();
    console.log(`Create your first admin user:`);
    const firstUser = await inquirer_1.default.prompt([
        {
            type: 'input',
            name: 'email',
            message: 'Email',
            default: 'admin@example.com',
        },
        {
            type: 'password',
            name: 'password',
            message: 'Password',
            mask: '*',
            validate: (input) => {
                if (input === null || input === '')
                    throw new Error('The password cannot be empty!');
                return true;
            },
        },
    ]);
    firstUser.password = await argon2_1.default.hash(firstUser.password);
    const userID = uuid_1.v4();
    const roleID = uuid_1.v4();
    await db('directus_roles').insert({
        id: roleID,
        name: 'Administrator',
        icon: 'verified',
        admin_access: true,
        description: 'Initial administrative role with unrestricted App/API access',
    });
    await db('directus_users').insert({
        id: userID,
        status: 'active',
        email: firstUser.email,
        password: firstUser.password,
        first_name: 'Admin',
        last_name: 'User',
        role: roleID,
    });
    await db.destroy();
    console.log(`
Your project has been created at ${chalk_1.default.green(rootPath)}.

The configuration can be found in ${chalk_1.default.green(rootPath + '/.env')}

Start Directus by running:
  ${chalk_1.default.blue('cd')} ${rootPath}
  ${chalk_1.default.blue('npx directus')} start
`);
    process.exit(0);
}
exports.default = init;
